File: blog/20121104/continuation_monad/callcc_sample2.hs

import Control.Monad.Cont

sample :: Int -> Cont r Int
sample n = callCC $ \cc1 -> do
	when (odd n) $ do
		-- (1)
		cc1 n

	r <- callCC $ \cc2 -> do
		when (n < 4) $ do
			-- (2)
			cc2 $ n * 1000

		when (n == 4) $ do
			-- (3)
			cc1 $ n * 100

		-- (4)
		return $ n * 10

	-- (5)
	return $ r + 1

main = do
	runCont (sample 1) print -- (1)
	runCont (sample 2) print -- (2) (5)
	runCont (sample 3) print -- (1)
	runCont (sample 4) print -- (3)
	runCont (sample 5) print -- (1)
	runCont (sample 6) print -- (4) (5)

File: scala/scalaz/continuation/CallCCSample2.scala
package fits.sample

import scalaz._
import Scalaz._

object CallCCSample2 extends App {
	import Continuation._

	type CtInt[R] = Continuation[R, Int]

	def sample[R](n: Int): CtInt[R] = callCC { cc1: (Int => CtInt[R]) =>

		if(n % 2 == 1) {
			cc1(n)
		}
		else {
			for {
				x <- callCC { cc2: (Int => CtInt[R]) => 
					if (n < 4) {
						cc2(n * 1000)
					}
					else if (n == 4) {
						cc1(n * 100)
					}
					else {
						continuationInstance[R].point(n * 10)
					}
				}
			} yield (x + 1)
		}
	}

	sample(1).runCont { println }
	sample(2).runCont { println }
	sample(3).runCont { println }
	sample(4).runCont { println }
	sample(5).runCont { println }
	sample(6).runCont { println }
}
